lib/rpcenc/reader.go

```diff
@@ -406,6 +406,29 @@ func ReaderParamDecoder() (http.HandlerFunc, jsonrpc.ServerOption) {
 			return
 		case <-req.Context().Done():
 			log.Errorf("context error in reader stream handler (2): %v", req.Context().Err())
+
+			closed := make(chan struct{})
+			// start a draining goroutine
+			go func() {
+				for {
+					select {
+					case r, ok := <-wr.res:
+						if !ok {
+							return
+						}
+						log.Errorw("discarding read res", "type", r.rt, "meta", r.meta)
+					case <-closed:
+						return
+					}
+				}
+			}()
```

Comment:
What does this extra `closed` channel do when we are immediately closing it? Isn't this going to race between the two `select` cases and possibly immediately returning without reading anything off of `wr.res`?

Reply:
It cleans up this goroutine.

`wr.res` can only be written in `wr.beginOnce`, and is only written to once, so by calling it below, we know that `wr.res` must have been written to already. 

This bit of code spins up a go-routine which will read pending sends into `wr.res`, then will make sure that `wr.beginOnce` was done (so `wr.res` can't be written to anymore), then it will shutdown the draining goroutine by closing the `closed` channel.